### DISCUSSION TITLE
Issue with Complex Multi-Join Query

### DISCUSSION THREAD

**Discussion by mxcdh** [prisma/prisma#19148] [Q&A]
Hello, 

I have a problem with multi-join queries.
I thinking, and testing many solution, but I waiting for result +1h and nothing.
Anyone have idea, how I can fix it, maybe I should change my database structure? 

- query
https://gist.github.com/mxcdh/d4ea7f98236e4a6ad084b9d1a2eaa7d9#file-postgres_query-sql

- schema.prisma
https://gist.github.com/mxcdh/d4ea7f98236e4a6ad084b9d1a2eaa7d9#file-schema-prisma

- what I want returned:
https://gist.github.com/mxcdh/d4ea7f98236e4a6ad084b9d1a2eaa7d9#file-return-md

- database table overview
https://gist.github.com/mxcdh/d4ea7f98236e4a6ad084b9d1a2eaa7d9#file-database_overview-md

**Comment by nurul3101** [comment]
Hello @mxcdh 👋 

What exact problem do you face with the raw SQL query you have shared?

> but I waiting for result +1h and nothing.

No records are returned for the SQL query?

**Reply by mxcdh** [reply]
@nurul3101  I running this query and waiting +1 hours and still waiting and calculate
![screenshot-2023-05-09 at 13 07 46@2x](https://github.com/prisma/prisma/assets/98115983/de007264-ccbd-4779-ae94-882877d821cf)
How speed up this query?

**Reply by nurul3101** [reply]
Can you get the query execution plan for the query by using [Explain Analyze](https://www.postgresql.org/docs/current/sql-explain.html) and check which database operation is taking the most time?
Creating an Index on the where clause fields would also help in speeding up the query.

**Reply by mxcdh** [reply]
@nurul3101  
I have @index in my prisma, right?
https://gist.github.com/mxcdh/d4ea7f98236e4a6ad084b9d1a2eaa7d9#file-schema-prisma

About EA, I added, but I still can wait 2-3 hours, and nothing:
https://dsh.re/c98928
Maybe I should wrong EA? 

btw
I made:https://www.db-fiddle.com/f/4jyoMCicNSZpjMt4jFYoz5/8558
it's look right, the model works well and fast on a small amount of data, but not on the whole database. 
What should I do? Should I change the structure?

**Reply by mxcdh** [reply]
@nurul3101 can you help me?

**Reply by nurul3101** [reply]
Can you try with just EXPLAIN and check if you get an execution plan?
```sql
EXPLAIN WITH RECURSIVE category_hierarchy AS (
  SELECT
    c.id,
    c.url,
    cd.name,
    cc.category_children_id,
    cc.category_parent_id,
    1 AS depth
  FROM
    category c
    JOIN category_children_to_category_parent cc ON c.id = cc.category_parent_id
    JOIN category_data cd ON c.id = cd.main_category_id
  WHERE
    cd.worker_id = (
      SELECT
        MAX(id)
      FROM
        worker
    )
  UNION ALL
  SELECT
    c.id,
    c.url,
    cd.name,
    cc.category_children_id,
    cc.category_parent_id,
    ch.depth + 1
  FROM
    category c
    JOIN category_children_to_category_parent cc ON c.id = cc.category_parent_id
    JOIN category_data cd ON c.id = cd.main_category_id
    JOIN category_hierarchy ch ON cc.category_parent_id = ch.category_children_id
)
SELECT
  id,
  url,
  name,
  depth
FROM
  category_hierarchy
ORDER BY
  depth,
  id;
```

**Reply by nurul3101** [reply]
Checking if using `EXPLAIN` worked for you?

**Reply by nurul3101** [reply]
I'll mark this discussion as closed. Please feel free to reopen if needed.
